File: HRForz-API-Node/src/core/security/jwt.ts
import * as jose from 'jose';
import { v4 as uuidv4 } from 'uuid';
import { config } from '../../config';
import { TokenExpired, TokenInvalid } from '../exceptions/base';

interface AccessTokenPayload {
  sub: string;
  employee_id: string;
  role: string;
  email: string;
  type: 'access';
}

interface RefreshTokenPayload {
  sub: string;
  type: 'refresh';
}

let privateKey: jose.KeyLike | null = null;
let publicKey: jose.KeyLike | null = null;

async function getPrivateKey(): Promise<jose.KeyLike> {
  if (!privateKey) {
    const pemKey = config.JWT_PRIVATE_KEY.replace(/\\n/g, '\n');
    privateKey = await jose.importPKCS8(pemKey, config.JWT_ALGORITHM);
  }
  return privateKey;
}

async function getPublicKey(): Promise<jose.KeyLike> {
  if (!publicKey) {
    const pemKey = config.JWT_PUBLIC_KEY.replace(/\\n/g, '\n');
    publicKey = await jose.importSPKI(pemKey, config.JWT_ALGORITHM);
  }
  return publicKey;
}

export async function createAccessToken(
  userId: string,
  employeeId: string,
  role: string,
  email: string
): Promise<string> {
  const key = await getPrivateKey();
  const now = Math.floor(Date.now() / 1000);
  const exp = now + config.ACCESS_TOKEN_EXPIRE_MINUTES * 60;

  return new jose.SignJWT({
    sub: userId,
    employee_id: employeeId,
    role,
    email,
    type: 'access',
    jti: uuidv4(),
  })
    .setProtectedHeader({ alg: config.JWT_ALGORITHM })
    .setIssuedAt(now)
    .setExpirationTime(exp)
    .sign(key);
}

export async function createRefreshToken(userId: string): Promise<string> {
  const key = await getPrivateKey();
  const now = Math.floor(Date.now() / 1000);
  const exp = now + config.REFRESH_TOKEN_EXPIRE_DAYS * 24 * 60 * 60;

  return new jose.SignJWT({
    sub: userId,
    type: 'refresh',
    jti: uuidv4(),
  })
    .setProtectedHeader({ alg: config.JWT_ALGORITHM })
    .setIssuedAt(now)
    .setExpirationTime(exp)
    .sign(key);
}

export async function decodeAccessToken(token: string): Promise<AccessTokenPayload> {
  try {
    const key = await getPublicKey();
    const { payload } = await jose.jwtVerify(token, key, {
      algorithms: [config.JWT_ALGORITHM],
    });

    if (payload.type !== 'access') {
      throw new TokenInvalid();
    }

    return payload as unknown as AccessTokenPayload;
  } catch (err) {
    if (err instanceof jose.errors.JWTExpired) {
      throw new TokenExpired();
    }
    if (err instanceof TokenInvalid || err instanceof TokenExpired) {
      throw err;
    }
    throw new TokenInvalid();
  }
}

export async function decodeRefreshToken(token: string): Promise<RefreshTokenPayload> {
  try {
    const key = await getPublicKey();
    const { payload } = await jose.jwtVerify(token, key, {
      algorithms: [config.JWT_ALGORITHM],
    });

    if (payload.type !== 'refresh') {
      throw new TokenInvalid();
    }

    return payload as unknown as RefreshTokenPayload;
  } catch (err) {
    if (err instanceof jose.errors.JWTExpired) {
      throw new TokenExpired();
    }
    if (err instanceof TokenInvalid || err instanceof TokenExpired) {
      throw err;
    }
    throw new TokenInvalid();
  }
}

File: HRForz-API/app/core/security/jwt.py
import uuid
from datetime import datetime, timedelta, timezone
from typing import Any

from jose import JWTError, jwt

from app.config import settings
from app.core.exceptions.base import TokenExpired, TokenInvalid


def _now_utc() -> datetime:
    return datetime.now(timezone.utc)


def create_access_token(
    user_id: uuid.UUID,
    employee_id: uuid.UUID,
    role: str,
    email: str,
) -> str:
    expire = _now_utc() + timedelta(minutes=settings.ACCESS_TOKEN_EXPIRE_MINUTES)
    payload: dict[str, Any] = {
        "sub": str(user_id),
        "employee_id": str(employee_id),
        "role": role,
        "email": email,
        "type": "access",
        "iat": _now_utc(),
        "exp": expire,
        "jti": str(uuid.uuid4()),
    }
    return jwt.encode(payload, settings.JWT_PRIVATE_KEY, algorithm=settings.JWT_ALGORITHM)


def create_refresh_token(user_id: uuid.UUID) -> str:
    expire = _now_utc() + timedelta(days=settings.REFRESH_TOKEN_EXPIRE_DAYS)
    payload: dict[str, Any] = {
        "sub": str(user_id),
        "type": "refresh",
        "iat": _now_utc(),
        "exp": expire,
        "jti": str(uuid.uuid4()),
    }
    return jwt.encode(payload, settings.JWT_PRIVATE_KEY, algorithm=settings.JWT_ALGORITHM)


def decode_access_token(token: str) -> dict[str, Any]:
    try:
        payload = jwt.decode(
            token,
            settings.JWT_PUBLIC_KEY,
            algorithms=[settings.JWT_ALGORITHM],
        )
        if payload.get("type") != "access":
            raise TokenInvalid()
        return payload
    except JWTError as exc:
        if "expired" in str(exc).lower():
            raise TokenExpired()
        raise TokenInvalid()


def decode_refresh_token(token: str) -> dict[str, Any]:
    try:
        payload = jwt.decode(
            token,
            settings.JWT_PUBLIC_KEY,
            algorithms=[settings.JWT_ALGORITHM],
        )
        if payload.get("type") != "refresh":
            raise TokenInvalid()
        return payload
    except JWTError as exc:
        if "expired" in str(exc).lower():
            raise TokenExpired()
        raise TokenInvalid()
